Sources/GRPC/HTTP1ToGRPCServerCodec.swift

```diff
@@ -394,14 +397,13 @@ extension HTTP1ToGRPCServerCodec: ChannelOutboundHandler {
 
         // TODO: Binary responses that are non multiples of 3 will end = or == when encoded in
         // base64. Investigate whether this might have any effect on the transport mechanism and
-        // client decoding. Initial results say that they are inocuous, but we might have to keep
+        // client decoding. Initial results say that they are innocuous, but we might have to keep
         // an eye on this in case something trips up.
-        if let binaryData = responseTextBuffer.readData(length: responseTextBuffer.readableBytes) {
-          let encodedData = binaryData.base64EncodedString()
-          responseTextBuffer.clear()
-          responseTextBuffer.reserveCapacity(encodedData.utf8.count)
-          responseTextBuffer.writeString(encodedData)
-        }
+        let binaryData = responseTextBuffer.readData(length: responseTextBuffer.readableBytes)!
+        let encodedData = binaryData.base64EncodedString()
```

Comment:
I was wondering why we had to do all that compaction of `ByteBuffer`s into a linear data structure, but now I see why. This is actually quite a lot of intermediate byte copying to only have to (potentially) copy those bytes again. At the absolute minimum we should use the flag to `readData` that prevents the copy, and hope that we will dodge the CoW operation that we hit here.

But the other option is to just do all of this work in `Data`-space. Given that we need a `Data` _anyway_, should we just copy all the bytes into a `Data` via `ByteBufferView`?

This will probably be marginally slower than doing this in `ByteBuffer`-land, but I suspect that slowness will be marginal at best, and we'll avoid a lot of complexity in _this_ code. We can still keep hold of the one buffer for clearing.

Reply:
Yeah, the gRPC Web path is a bit unfortunate... The `Data`-land approach makes sense, I'll do that.